File: apps/workbench/lib/bff/admin/knowledge.ts
// KnowledgeOps slot handlers for the Admin BFF (ADR-0087 master-detail policy
// authoring over the six Required Operational Policy Slots, ADR-0003). Pure and
// dependency-injected; the thin app/api/admin/knowledge route files wrap these
// with withSession and inject the real KnowledgeStore singleton.
import { HermesApiClient, HermesApiError } from "../../gateway/hermes-api-client";
import { hermesErrorToProblem } from "../../gateway/hermes-error";
import type { PolicySlot, SlotStatus } from "../../gateway/knowledge-store";
import { json, problem } from "../respond";
import { type AdminDeps, readJsonBody } from "./deps";

export function handleListSlots(deps: AdminDeps): Response {
  return json({ slots: deps.knowledge.listSlots() });
}

export async function handleSaveDraft(
  req: Request,
  slotId: string,
  deps: AdminDeps,
): Promise<Response> {
  const body = await readJsonBody(req);
  const patch: { draftText?: string; owner?: string; reviewDate?: string } = {};
  if (typeof body?.draftText === "string") patch.draftText = body.draftText;
  if (typeof body?.owner === "string") patch.owner = body.owner;
  if (typeof body?.reviewDate === "string") patch.reviewDate = body.reviewDate;

  const slot = deps.knowledge.saveDraft(slotId, patch);
  if (!slot) return problem(404, "slot not found");
  return json({ slot });
}

export function handleSubmitSlot(slotId: string, deps: AdminDeps): Response {
  const result = deps.knowledge.submitForEval(slotId);
  if (result.ok) return json({ slot: result.slot });
  if (result.reason === "not_found") return problem(404, "slot not found");
  return problem(409, "slot has no draft to submit");
}

export function handleRollbackSlot(slotId: string, deps: AdminDeps): Response {
  const result = deps.knowledge.rollbackPublished(slotId);
  if (result.ok) return json({ slot: result.slot });
  if (result.reason === "not_found") return problem(404, "slot not found");
  return problem(409, "slot has no previous published version");
}

// --- Per-profile API cutover (ADR-0141/0145 Increment 6) ---------------------
// The Supervisor Admin knowledge-slot routes dispatch toee_knowledge_ops over the
// per-profile Hermes API when HERMES_ADMIN_API_URL/TOKEN are configured (else the
// in-memory KnowledgeStore). The list read uses dispatch (fail-open); the three
// governed mutations use dispatchWrite (fail-closed on the acting supervisor baked
// into the client), so a write can never land a NULL-actor audit row — the
// datastore enforces the same rule. The per-class error mapping turns a governed
// not_found/conflict into 404/409 (store-path status parity).

const SLOT_STATUSES = new Set<SlotStatus>([
  "empty",
  "draft",
  "pending_eval",
  "published",
  "gap",
]);

// Maps a snake_case workbench_policy_slot row from the per-profile API (ADR-0145)
// onto the wire PolicySlot, rejecting contract violations (missing id/title,
// unknown status) as governed HermesApiErrors so a bad upstream surfaces on the
// ADR-0090 banner rather than rendering garbage (mirrors mapPublicAccount).
export function mapPolicySlot(raw: unknown): PolicySlot {
  if (typeof raw !== "object" || raw === null) {
    throw new HermesApiError("unexpected_error", "malformed policy slot payload");
  }
  const r = raw as Record<string, unknown>;
  const slotId = typeof r.slot_id === "string" ? r.slot_id : "";
  if (!slotId) throw new HermesApiError("unexpected_error", "missing slot id");
  if (typeof r.title !== "string" || r.title.length === 0) {
    throw new HermesApiError("unexpected_error", "missing slot title");
  }
  if (typeof r.status !== "string" || !SLOT_STATUSES.has(r.status as SlotStatus)) {
    throw new HermesApiError(
      "unexpected_error",
      `unknown slot status: ${String(r.status)}`,
    );
  }
  // draft_text/published_text/owner/review_date are nullable; a provided string
  // (incl. "") is preserved, anything else is null — store-path parity.
  const orNull = (v: unknown): string | null => (typeof v === "string" ? v : null);
  return {
    slotId,
    title: r.title,
    status: r.status as SlotStatus,
    draftText: orNull(r.draft_text),
    publishedText: orNull(r.published_text),
    owner: orNull(r.owner),
    reviewDate: orNull(r.review_date),
    hasGapPrompt: r.has_gap_prompt === true,
  };
}

export async function handleListSlotsViaApi(
  client: HermesApiClient,
): Promise<Response> {
  try {
    const data = (await client.dispatch(
      "toee_knowledge_ops",
      "get_policy_slots",
    )) as { slots?: unknown };
    const rows = Array.isArray(data?.slots) ? data.slots : [];
    return json({ slots: rows.map(mapPolicySlot) });
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export async function handleSaveDraftViaApi(
  req: Request,
  slotId: string,
  client: HermesApiClient,
): Promise<Response> {
  const body = await readJsonBody(req);
  // Same provided-field semantics as the store path: only forward present fields,
  // mapped to the snake_case the datastore reads.
  const params: Record<string, unknown> = { slot_id: slotId };
  if (typeof body?.draftText === "string") params.draft_text = body.draftText;
  if (typeof body?.owner === "string") params.owner = body.owner;
  if (typeof body?.reviewDate === "string") params.review_date = body.reviewDate;
  try {
    const result = (await client.dispatchWrite(
      "toee_knowledge_ops",
      "update_policy_slot",
      params,
    )) as { slot?: unknown };
    return json({ slot: mapPolicySlot(result.slot) });
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export async function handleSubmitSlotViaApi(
  slotId: string,
  client: HermesApiClient,
): Promise<Response> {
  try {
    const result = (await client.dispatchWrite(
      "toee_knowledge_ops",
      "submit_for_eval",
      { slot_id: slotId },
    )) as { slot?: unknown };
    return json({ slot: mapPolicySlot(result.slot) });
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export async function handleRollbackSlotViaApi(
  slotId: string,
  client: HermesApiClient,
): Promise<Response> {
  try {
    const result = (await client.dispatchWrite(
      "toee_knowledge_ops",
      "rollback_published_policy",
      { slot_id: slotId },
    )) as { slot?: unknown };
    return json({ slot: mapPolicySlot(result.slot) });
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

// --- S11: corpus status + retrieval probe (FR-6) ------------------------------
// Both routes are API-only (no in-memory store fallback -- there is no corpus to
// fake): the thin app/api route returns 503 up front when the admin API client
// isn't configured (see createAdminApiClient), so these handlers can assume a
// real client.

// 0.0.4 S04 (FR-11): the re-ingest panel's status readback. Null when no
// re-ingest has ever been queued, or on a backend with no durable queue.
export type IngestJobStatus = {
  jobId: string;
  status: string;
  attempts: number;
  lastError: string | null;
  queuedAt: string | null;
  updatedAt: string | null;
};

export type CorpusStatus = {
  docCount: number;
  chunkCount: number;
  lastIngestAt: string | null;
  byType: { pageType: string; count: number }[];
  lastIngestJob: IngestJobStatus | null;
};

// Maps the snake_case toee_knowledge_ops.get_corpus_status payload onto the wire
// CorpusStatus, rejecting a malformed upstream as a governed HermesApiError
// (mirrors mapPolicySlot).
export function mapCorpusStatus(raw: unknown): CorpusStatus {
  if (typeof raw !== "object" || raw === null) {
    throw new HermesApiError("unexpected_error", "malformed corpus status payload");
  }
  const r = raw as Record<string, unknown>;
  const byType = Array.isArray(r.by_type) ? r.by_type : [];
  return {
    docCount: typeof r.doc_count === "number" ? r.doc_count : 0,
    chunkCount: typeof r.chunk_count === "number" ? r.chunk_count : 0,
    lastIngestAt: typeof r.last_ingest_at === "string" ? r.last_ingest_at : null,
    byType: byType.map((row) => {
      const rr = row as Record<string, unknown>;
      return {
        pageType: typeof rr.page_type === "string" ? rr.page_type : "",
        count: typeof rr.count === "number" ? rr.count : 0,
      };
    }),
    lastIngestJob: mapIngestJob(r.last_ingest_job),
  };
}

function mapIngestJob(raw: unknown): IngestJobStatus | null {
  if (typeof raw !== "object" || raw === null) return null;
  const r = raw as Record<string, unknown>;
  if (typeof r.job_id !== "string" || typeof r.status !== "string") return null;
  const orNull = (v: unknown): string | null => (typeof v === "string" ? v : null);
  return {
    jobId: r.job_id,
    status: r.status,
    attempts: typeof r.attempts === "number" ? r.attempts : 0,
    lastError: orNull(r.last_error),
    queuedAt: orNull(r.queued_at),
    updatedAt: orNull(r.updated_at),
  };
}

export type ReingestQueued = { jobId: string | null; status: string };

// S04 (FR-11): the re-ingest trigger 0.0.3 S11 shipped as a display-only stub.
// A governed dispatchWrite -- it TRUNCATEs and reloads the whole corpus, so it is
// fail-closed on the acting supervisor and audited on the Hermes side; the panel
// then reads the job back through get_corpus_status.
export async function handleTriggerReingestViaApi(
  client: HermesApiClient,
): Promise<Response> {
  try {
    const data = (await client.dispatchWrite(
      "toee_knowledge_ops",
      "enqueue_corpus_reingest",
      {},
    )) as Record<string, unknown>;
    const jobId = typeof data?.job_id === "string" ? data.job_id : null;
    const status = typeof data?.status === "string" ? data.status : "queued";
    return json({ jobId, status } satisfies ReingestQueued);
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export async function handleGetCorpusStatusViaApi(
  client: HermesApiClient,
): Promise<Response> {
  try {
    const data = await client.dispatch("toee_knowledge_ops", "get_corpus_status");
    return json({ status: mapCorpusStatus(data) });
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export type ProbeResult = {
  title: string;
  url: string | null;
  snippet: string;
};

function mapProbeResult(raw: unknown): ProbeResult {
  const r = (raw && typeof raw === "object" ? raw : {}) as Record<string, unknown>;
  return {
    title: typeof r.title === "string" ? r.title : "",
    url: typeof r.url === "string" ? r.url : null,
    snippet: typeof r.snippet === "string" ? r.snippet : "",
  };
}

// Retrieval probe (S08's re-scoped layer-② evidence, per S11's scope addition):
// dispatches toee_knowledge_search.search_public_site through the admin profile
// API, hitting the REAL S09/S10 driver when the dispatch server runs with
// KNOWLEDGE_BACKEND=retriever. A blank query is rejected before dispatch (mirrors
// the KnowledgeDriver's own empty-query guard, but as a normal 400 -- there's no
// governed tool call to make yet).
export async function handleProbeQueryViaApi(
  req: Request,
  client: HermesApiClient,
): Promise<Response> {
  const body = await readJsonBody(req);
  const query = typeof body?.query === "string" ? body.query : "";
  if (!query.trim()) return problem(400, "query is required");
  try {
    const data = (await client.dispatch("toee_knowledge_search", "search_public_site", {
      query,
    })) as { results?: unknown };
    const rows = Array.isArray(data?.results) ? data.results : [];
    return json({ results: rows.map(mapProbeResult) });
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

File: hermes-runtime/hermes_runtime/datastore/handlers/knowledge.py
"""Datastore handlers for ``toee_knowledge_ops`` (ADR-0003/0087/0145).

The Supervisor Admin ``/admin/knowledge`` master-detail (ADR-0087) authors the six
Required Operational Policy Slots (ADR-0003) with a draft -> pending_eval ->
published lifecycle, separate draft/published text, owner/review metadata, and a
published-version history for rollback. ADR-0145 cuts this over from the in-memory
``KnowledgeStore`` (the behavioral source of truth) onto a dedicated
``workbench_policy_slot`` (+ ``workbench_policy_slot_history``) table that mirrors
its ``PolicySlot`` shape field-for-field. Writes are Supervisor governance: each is
fail-closed on an attributed actor and appends a Workbench Audit Log row.

ADR-0145 left ``submit`` decoupled from any publish step; ADR-0146 connects them:
promoting a ``policy_publish`` eval run now publishes the matching authoring slot
(``eval_review._publish_authoring_slot``), pushing the prior ``published_text`` onto
``workbench_policy_slot_history`` so :func:`_rollback_published_policy` has data. The
old ``knowledge_version`` publish path is retired (promotion is runId-keyed now).
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from psycopg.rows import dict_row

from toee_hermes.errors import ToolDriverError

from ._common import insert_audit, read_string, serialize_row

if TYPE_CHECKING:  # pragma: no cover - typing only
    from toee_hermes.tool_gate import ToolExecutionContext

# The wire-safe PolicySlot read model the BFF maps onto camelCase (ADR-0145).
_SLOT_COLUMNS = (
    "slot_id, title, status, draft_text, published_text, owner, review_date, has_gap_prompt"
)


def _require_actor(context: "ToolExecutionContext") -> str:
    """The acting Supervisor Admin for a governed knowledge write, or a denial.

    KnowledgeOps mutations are actor-attributed governance (ADR-0029/0087/0141):
    the actor rides ``ToolExecutionContext.user_id``, asserted by the BFF under the
    shared bearer. Trusting an absent actor would write the mutation *and* a
    NULL-actor audit row while returning success. Fail closed so every knowledge
    write handler is protected at once. (Mirrors ``accounts._require_actor``.)
    """
    actor = context.user_id
    if not actor:
        raise ToolDriverError(
            "policy_blocked", "A governed knowledge write requires an attributed actor."
        )
    return actor


def _optional_str(params: dict[str, Any], *keys: str) -> Optional[str]:
    """First *provided* string among ``keys`` (snake-first), keeping ``""``.

    Unlike ``read_string`` (which drops empty strings), this distinguishes an
    explicitly-provided empty string from an absent key, so ``saveDraft`` parity
    holds: sending ``draftText: ""`` clears the draft, while omitting it leaves the
    stored value untouched (``COALESCE(NULL, col)``).
    """
    for key in keys:
        value = params.get(key)
        if isinstance(value, str):
            return value
    return None


def _slot_row(conn, slot_id: str) -> Any:
    with conn.cursor(row_factory=dict_row) as cur:
        cur.execute(
            f"SELECT {_SLOT_COLUMNS} FROM workbench_policy_slot WHERE slot_id = %s",
            (slot_id,),
        )
        return serialize_row(cur.fetchone())


def _get_policy_slots(conn, params: dict[str, Any], context: "ToolExecutionContext") -> Any:
    # listSlots() parity: every Required Operational Policy Slot, in the fixed
    # ADR-0003 list order (sort_order). A read -> no actor required, no audit.
    with conn.cursor(row_factory=dict_row) as cur:
        cur.execute(
            f"SELECT {_SLOT_COLUMNS} FROM workbench_policy_slot ORDER BY sort_order"
        )
        rows = cur.fetchall()
    return {"slots": [serialize_row(row) for row in rows]}


def _update_policy_slot(conn, params: dict[str, Any], context: "ToolExecutionContext") -> Any:
    # saveDraft parity: patch only the provided fields and flip empty/gap -> draft
    # when the resulting draft text is non-empty, atomically in one UPDATE. An
    # unknown slot id is not_found (BFF -> 404), store-path parity.
    actor = _require_actor(context)
    slot_id = read_string(params, "slot_id", "slotId", "slot")
    if slot_id is None:
        raise ToolDriverError("unexpected_error", "slot_id is required.")
    fields = {
        "slot_id": slot_id,
        "draft_text": _optional_str(params, "draft_text", "draftText"),
        "owner": _optional_str(params, "owner"),
        "review_date": _optional_str(params, "review_date", "reviewDate"),
    }
    with conn.cursor(row_factory=dict_row) as cur:
        cur.execute(
            f"""
            UPDATE workbench_policy_slot SET
                draft_text = COALESCE(%(draft_text)s, draft_text),
                owner = COALESCE(%(owner)s, owner),
                review_date = COALESCE(%(review_date)s, review_date),
                status = CASE
                    WHEN COALESCE(%(draft_text)s, draft_text) IS NOT NULL
                         AND length(COALESCE(%(draft_text)s, draft_text)) > 0
                         AND status IN ('empty', 'gap')
                    THEN 'draft'
                    ELSE status
                END,
                updated_at = now()
            WHERE slot_id = %(slot_id)s
            RETURNING {_SLOT_COLUMNS}
            """,
            fields,
        )
        row = cur.fetchone()
    if row is None:
        raise ToolDriverError("not_found", "slot not found")
    insert_audit(
        conn,
        profile=context.profile,
        account_id=actor,
        action="update_policy_slot",
        target_type="policy_slot",
        target_id=slot_id,
        details={"status": row["status"]},
    )
    return {"slot": serialize_row(row)}


def _submit_for_eval(conn, params: dict[str, Any], context: "ToolExecutionContext") -> Any:
    # submitForEval parity: requires a non-empty draft. The atomic UPDATE only
    # matches a slot that has one; rowcount 0 is disambiguated into not_found (404,
    # unknown slot) vs conflict (409, "slot has no draft to submit").
    actor = _require_actor(context)
    slot_id = read_string(params, "slot_id", "slotId", "slot")
    if slot_id is None:
        raise ToolDriverError("unexpected_error", "slot_id is required.")
    with conn.cursor(row_factory=dict_row) as cur:
        cur.execute(
            f"""
            UPDATE workbench_policy_slot
            SET status = 'pending_eval', updated_at = now()
            WHERE slot_id = %s AND draft_text IS NOT NULL AND length(draft_text) > 0
            RETURNING {_SLOT_COLUMNS}
            """,
            (slot_id,),
        )
        row = cur.fetchone()
    if row is None:
        with conn.cursor() as cur:
            cur.execute("SELECT 1 FROM workbench_policy_slot WHERE slot_id = %s", (slot_id,))
            exists = cur.fetchone() is not None
        if not exists:
            raise ToolDriverError("not_found", "slot not found")
        raise ToolDriverError("conflict", "slot has no draft to submit")
    insert_audit(
        conn,
        profile=context.profile,
        account_id=actor,
        action="submit_for_eval",
        target_type="policy_slot",
        target_id=slot_id,
        details={},
    )
    return {"slot": serialize_row(row)}


def _rollback_published_policy(
    conn, params: dict[str, Any], context: "ToolExecutionContext"
) -> Any:
    # rollbackPublished parity: restore the previous published text from history.
    # Unknown slot -> not_found (404); no prior version -> conflict (409). The pop
    # is a single DELETE ... RETURNING so two concurrent rollbacks can't both claim
    # the same history row.
    actor = _require_actor(context)
    slot_id = read_string(params, "slot_id", "slotId", "slot")
    if slot_id is None:
        raise ToolDriverError("unexpected_error", "slot_id is required.")
    with conn.cursor() as cur:
        cur.execute("SELECT 1 FROM workbench_policy_slot WHERE slot_id = %s", (slot_id,))
        if cur.fetchone() is None:
            raise ToolDriverError("not_found", "slot not found")
    with conn.cursor() as cur:
        cur.execute(
            """
            DELETE FROM workbench_policy_slot_history
            WHERE id = (
                SELECT id FROM workbench_policy_slot_history
                WHERE slot_id = %s
                ORDER BY archived_at DESC, id DESC
                LIMIT 1
            )
            RETURNING published_text
            """,
            (slot_id,),
        )
        popped = cur.fetchone()
    if popped is None:
        raise ToolDriverError("conflict", "slot has no previous published version")
    with conn.cursor() as cur:
        cur.execute(
            "UPDATE workbench_policy_slot"
            " SET published_text = %s, status = 'published', updated_at = now()"
            " WHERE slot_id = %s",
            (popped[0], slot_id),
        )
    insert_audit(
        conn,
        profile=context.profile,
        account_id=actor,
        action="rollback_published_policy",
        target_type="policy_slot",
        target_id=slot_id,
        details={},
    )
    return {"slot": _slot_row(conn, slot_id)}


# The eval publish-gate that promotion now drives lives on workbench_policy_slot
# (ADR-0146): a promoted policy_publish eval run publishes the matching authoring
# slot via eval_review._publish_authoring_slot, pushing the prior published text to
# workbench_policy_slot_history. The old knowledge_version publish path
# (publish_pending_slot/_latest_version/_require_known_slot) is retired — promotion
# is runId-keyed now (ADR-0146 resolves ADR-0145 divergences #1/#2).


# --- corpus status (S11, FR-6) -----------------------------------------------
# Unlike every write above, this read is CROSS-DATABASE: workbench_policy_slot
# lives on the business datastore (the ``conn`` the registry hands the handler),
# but the corpus itself (``knowledge_chunk``) lives on the SEPARATE toee_knowledge
# database (S-ISO isolation invariant, hermes_runtime/knowledge/config.py). So
# _get_corpus_status ignores the business ``conn`` and opens its own short-lived
# read-only connection to the knowledge DSN.


def _corpus_status_from_conn(kconn) -> Any:
    """Doc/chunk counts + last-ingest time + per-type breakdown from an OPEN
    connection to the knowledge database. Pure read (no writes, no commit) --
    the injectable core so tests can point it at a throwaway knowledge schema
    (mirrors the S08 retriever's ``retrieve(conn=...)`` seam)."""
    with kconn.cursor() as cur:
        cur.execute(
            "SELECT count(*), count(DISTINCT page_id), max(created_at) FROM knowledge_chunk"
        )
        chunk_count, doc_count, last_ingest_at = cur.fetchone()
        cur.execute(
            "SELECT page_type, count(*) FROM knowledge_chunk"
            " GROUP BY page_type ORDER BY page_type"
        )
        by_type = cur.fetchall()
    return {
        "doc_count": doc_count or 0,
        "chunk_count": chunk_count or 0,
        "last_ingest_at": last_ingest_at.isoformat() if last_ingest_at else None,
        "by_type": [{"page_type": pt, "count": n} for pt, n in by_type],
    }


def _last_ingest_job(conn) -> Any:
    """The most recent ``ingest`` job's status, for the panel's readback (S04).

    Unlike the corpus counts above this IS a business-database read -- the ``job``
    table lives there -- so it uses the handler's own ``conn``. ``None`` when no
    re-ingest has ever been queued, and ``None`` (not an error) on a database that
    predates migration 0011, so an un-migrated deployment still renders the panel.
    """
    # Local import: job_queue reaches back into datastore.handlers._common for
    # new_id, so a module-level import here would be a cycle.
    from hermes_runtime.job_queue import INGEST_JOB_TYPE

    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id, status, attempts, last_error, created_at, updated_at"
                " FROM job WHERE type = %s ORDER BY created_at DESC LIMIT 1",
                (INGEST_JOB_TYPE,),
            )
            row = cur.fetchone()
    except Exception:
        return None
    if row is None:
        return None
    job_id, status, attempts, last_error, created_at, updated_at = row
    return {
        "job_id": job_id,
        "status": status,
        "attempts": attempts,
        "last_error": last_error,
        "queued_at": created_at.isoformat() if created_at else None,
        "updated_at": updated_at.isoformat() if updated_at else None,
    }


def _get_corpus_status(conn, params: dict[str, Any], context: "ToolExecutionContext") -> Any:
    # A read -> no actor required, no audit (parity with the other knowledge
    # reads above).
    del params, context
    from hermes_runtime.knowledge.pool import get_knowledge_pool

    with get_knowledge_pool().connection() as kconn:
        status = _corpus_status_from_conn(kconn)
    # S04 (FR-11): the re-ingest panel's status readback. Cross-database again --
    # the counts came from toee_knowledge, the job row from the business `conn`.
    status["last_ingest_job"] = _last_ingest_job(conn) if conn is not None else None
    return status


def _enqueue_corpus_reingest(
    conn, params: dict[str, Any], context: "ToolExecutionContext"
) -> Any:
    """Queue an ``ingest`` job for the background worker (0.0.4 S04, FR-11).

    Replaces 0.0.3 S11's display-only stub (the panel printed a CLI command). A
    governed WRITE: it triggers a TRUNCATE-and-reload of the whole corpus, so it
    is dispatchWrite/fail-closed on the acting supervisor and writes a
    ``corpus_reingest_queued`` audit row -- the sibling precedent is
    ``trigger_retention_sweep``'s own audit row.

    ponytail: ``max_attempts=1``. Ingest is a heavy, non-idempotent-in-cost
    TRUNCATE + re-embed of the whole corpus; three automatic attempts would wipe
    and reload it three times over a transient fastembed/OOM failure. A failed
    ingest goes straight to ``dead`` for S05's governed replay instead. Raise it
    only if ingest ever becomes cheap and resumable **and** the lease is solved
    first: an ingest that outlives ``job_queue.DEFAULT_LEASE_SECONDS`` (300 s) is
    reclaimed mid-run, and that is harmless today ONLY because
    ``attempts >= max_attempts`` sends the reclaimed row to ``dead`` (never
    claimable) rather than back to ``failed``. Above 1, a reclaimed-but-still-
    running ingest becomes claimable again and two processes TRUNCATE the corpus
    concurrently -- so chunking ingest into cheap resumable pieces satisfies the
    first half of this condition while leaving that hazard fully intact.
    """
    del params
    from hermes_runtime.job_queue import INGEST_JOB_TYPE, insert_job

    _require_actor(context)
    with conn.cursor() as cur:
        job_id, _created = insert_job(
            cur,
            {"profile": context.profile, "actor_account_id": context.user_id},
            job_type=INGEST_JOB_TYPE,
            max_attempts=1,
        )
    insert_audit(
        conn,
        profile=context.profile,
        account_id=context.user_id,
        action="corpus_reingest_queued",
        target_type="knowledge_chunk",
        target_id=job_id,
        details={"job_id": job_id},
    )
    return {"job_id": job_id, "status": "queued"}


def knowledge_handlers() -> dict[str, dict[str, Any]]:
    """Registry fragment for the knowledge-ops datastore tool."""
    return {
        "toee_knowledge_ops": {
            "get_policy_slots": _get_policy_slots,
            "update_policy_slot": _update_policy_slot,
            "submit_for_eval": _submit_for_eval,
            "rollback_published_policy": _rollback_published_policy,
            "get_corpus_status": _get_corpus_status,
            "enqueue_corpus_reingest": _enqueue_corpus_reingest,
        }
    }
